src/Orleans.CodeGenerator/CodeGenerator.cs

```diff
@@ -492,21 +501,31 @@ private IEnumerable<IMemberDescription> GetDataMembers(INamedTypeSymbol symbol)
                 if (member is IPropertySymbol prop)
                 {
                     var id = GetId(prop);
+                    var isPrimaryConstructorParameter = false;
 
                     if (!id.HasValue)
                     {
-                        if (hasAttributes || !_options.GenerateFieldIds)
+                        if (hasAttributes || !allowImplicitFieldIds)
                         {
                             continue;
                         }
 
-                        id = ++nextFieldId;
+                        var primaryConstructorParameter = constructorParameters.FirstOrDefault(x => x.Name.Equals(prop.Name, StringComparison.OrdinalIgnoreCase));
+                        if (primaryConstructorParameter is not null)
+                        {
+                            id = (ushort)constructorParameters.IndexOf(primaryConstructorParameter);
+                            isPrimaryConstructorParameter = true;
+                        }
+                        else
+                        {
+                            id = ++nextFieldId;
```

Comment:
This logic is rather scary, since adding a ctor param implicitly shifts other fields.
Maybe field ids should be generated deterministically based upon some hash of their name as a way of avoiding this (and that logic would also need to be replicated in the analyzer which generates field ids)

Reply:
The code only runs when using implicit ids, where version tolerance is not really a thing (currently, anyway).

But I think the idea of using a hash to determine field ids is worthwhile, if we can use it to provide some level of predictable version tolerance even when using implicit field ids. We'd need a strategy for how to deal with hash collisions, and we may need to calculate the hash from a case-insensitive name. An alternative (or addition) to this could be to enforce ctor requirements (for annotated ctors, e.g. that parameters must match property declaration order) to prevent the most common mistakes. The latter could also be analyzer-only. 

Would it be reasonable to create a separate issue for this or would you like me to amend this PR?

Reply:
> We'd need a strategy for how to deal with hash collisions, and we may need to calculate the hash from a case-insensitive name. 

Failing to build is a good option in that rare case: the analyzer+codefix can add the computed id and have the developer resolve the conflict.